infinitic-worker-pulsar/src/main/kotlin/io/infinitic/worker/main/Application.kt

```diff
@@ -0,0 +1,142 @@
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import org.apache.pulsar.client.api.Message
+import org.apache.pulsar.client.api.PulsarClient
+import org.apache.pulsar.client.api.PulsarClientException
+import kotlin.concurrent.thread
+import kotlin.coroutines.CoroutineContext
+
+class Application internal constructor(private val pulsarClient: PulsarClient, private val worker: Worker) : CoroutineScope {
+    override val coroutineContext: CoroutineContext
+        get() = Dispatchers.IO + Job()
+
+    init {
+        Runtime.getRuntime().addShutdownHook(thread(start = false) {
+            println("Stopping because of the shutdown hook.")
+            stop()
+        })
+    }
+
+    fun run() {
+        try {
+            runWithConcurrency()
+        } catch (e: PulsarClientException) {
+            println(e) // FIXME: Remove and replace by a logger
+            stop()
+
+            throw e
+        }
+    }
+
+    fun runWithoutConcurrency() {
+        worker.getRegisteredTasks().map { name ->
+            pulsarClient.newTaskConsumer(name)
+        }.forEach { consumer ->
+            startConsumer(consumer) { message ->
+                message
+                    .let { AvroConverter.fromWorkers(it.value) }
+                    .let { worker.handle(it) }
+            }
+        }
+    }
+
+    fun runWithConcurrency() {
+        val messagesInputChannel = Channel<Message<AvroEnvelopeForWorker>>()
+        val workerInputChannel = Channel<MessageToProcess<ForWorkerMessage>>()
+
+        // launch 8 workers
+        repeat(8) {
+            launch(CoroutineName("worker-$it")) {
+                for (work in workerInputChannel) {
+                    worker.handle(work.message)
+                    work.replyTo.send(MessageProcessed(work.messageId))
+                }
+            }
+        }
+
+        worker.getRegisteredTasks().map { name ->
+            Pair(name, pulsarClient.newTaskConsumer(name))
+        }.forEach { (taskName, consumer) ->
+
+            // launch 1 consumer
+            launch(CoroutineName("consumer-$taskName")) {
+                while (isActive) {
+                    consumer
+                        .receiveSuspend()
+                        .let { messagesInputChannel.send(it) }
+                }
+            }
+
+            val resultChannel = Channel<MessageProcessed>()
+
+            // launch 1 acknowledger
+            launch(CoroutineName("ack-$taskName")) {
+                for (result in resultChannel) {
+                    val message = resultChannel.receive()
+                    consumer.acknowledgeSuspend(message.messageId)
+                }
+            }
+
+            // launch 1 executor
+            launch(CoroutineName("executor-$taskName")) {
```

Comment:
this could be avoided and put within consumer above

Reply:
fixed